sympy/ntheory/factor_.py

```diff
@@ -509,22 +509,56 @@ def perfect_power(n, candidates=None, big=True, factor=True):
             if pp:
                 pp = (n.func(1, pp[0]), pp[1])
         else:
-            pp = perfect_power(p)
-            if pp:
-                num, e = pp
-                pq = perfect_power(q, [e])
-                if pq:
-                    den, _ = pq
-                    pp = n.func(num, den), e
+            pp_num = perfect_power(p)
```

Comment:
Here and above, candidates, big and factor should be passed.

Reply:
You could call as `perfect_power(p, **kw)` if you capture `kw={'candidates': candidates, 'big': big, 'factor': factor}` at the start. But maybe big=True should be passed for the denominator.

Reply:
What about
```python
def perfect_power(n, candidates=None, big=True, factor=True):
    if n < 0:
        if candidates is None:
            pp = perfect_power(-n, big=True, factor=factor)
            if not pp:
                return False
            b, e = pp
            # if n is negative, the exponent must be odd
            e2 = e & (-e)
            b, e = b**e2, e//e2
            if e == 1:
                return False
            if big or isprime(e):
                return -b, e
            # report as e the smallest odd factor of e
            for p in primerange(3, e + 1):  # one p must succeed
                if e % p == 0:
                    return -b**(e//p), p
        odd_candidates = (i for i in candidates if i%2)
        if not odd_candidates:
            return False  # e must be odd but there were no odd candidates
        pp = perfect_power(-n, odd_candidates, big, factor)
        if pp:
            return -pp[0], pp[1]
        return False
    if isinstance(n, Rational) and not isinstance(n, Integer):
        p, q = n.p, n.q
        if p == 1:
            qq = perfect_power(q, candidates, big, factor)
            if qq is not None:
                return S.One/qq[0], qq[1]
            return False
        pp = perfect_power(p)
        if pp is None:
            return False
        qq = perfect_power(q)
        if qq is None:
            return False
        if candidates:
            candidates = [i for i in candidates if pp[1]%i == qq[1]%i == 0]
            if not candidates:
                return False
            e = max(candidates) if big else min(candidates)
            return n.func(pp[0]**(pp[1]//e),qq[0]**(qq[1]//e)), e
        g = math.gcd(pp[1], qq[1])
        if g == 1:
            return False
        if big:
            return n.func(pp[0]**(pp[1]//g), qq[0]**(qq[1]//g)), g
        # find first small candidate that divides g
        for e in primerange(2, g+1):
            if g%e == 0:
                break
        return n.func(pp[0]**(pp[1]//e), qq[0]**(qq[1]//e)), e
    # the current code for positive integers
```

Since `candidates` is advertised mainly as internal, I wish it could be removed since it is no longer needed in recursive calls -- as you show, dealing with the gcd is the way to go. But I think the above is a way to handle everything as advertised, e.g.

```
r = S(2)**(2*5*7)/S(3)**(2*7)
assert perfect_power(r) == (S(32)/3, 14)
assert perfect_power(-r) == (-S(1024)/9, 7)
assert perfect_power(r, big=False) == (S(34359738368)/2187, 2)
assert perfect_power(r, [2, 5]) == (S(34359738368)/2187, 2)
assert perfect_power(r, [5, 7]) == (S(1024)/9, 7)
assert perfect_power(r, [5, 7], big=False) == (S(1024)/9, 7)
assert perfect_power(r, [2, 5, 7], big=False) == (S(34359738368)/2187, 2)
assert perfect_power(-r, [5, 7], big=False) == (-S(1024)/9, 7)
```

Reply:
Absolutely. I was anticipating a version after review that incorporates the keywords. Thank you so much for this fully compatible approach. 

The base adjustment/tuple construction repeats a few times so I guess we can add a helper function to do that stuff. Something like:
```python
def compute_tuple(exponent):
    """Helper to compute final result given an exponent"""
    new_num = num_base ** (num_exp // exponent)
    new_den = den_base ** (den_exp // exponent)
    return n.func(new_num, new_den), exponent
```